[WIP] Optimize GC bit handling for small object pools

rainers opened the pull request:
This continues work of https://github.com/D-Programming-Language/druntime/pull/1127

The idea is to extract the actual bit handling into the different type of pools, so the small object pool can have a very different approach than the large object pool.

The actual change to the bit layout is in commit https://github.com/rainers/druntime/commit/42c133134974f2edd9a2e6ec4604624e6366f44a where the bits are stored in the same page as the objects they describe. For simplicity they are stored as one byte per object, saving space for objects larger than 16 bytes due to the removal of bit strides.

Unfortunately, there is no noticable performance improvement from this change yet. Intermediate version had a small loss caused by maintaining freebits instead of rebuilding them before a collection.

MartinNowak commented:
> bits are stored in the same page as the objects they describe.

This is no-go, because it kills COW. The Ruby GC made that mistake, try https://www.google.de/search?q=ruby%20gc%20cow.
Specializing the data structures for small and large objects is a good idea OTOH.
One this I wanted to try was using fixed size pools (4-16MB) for small and large objects and adding a huge size class for the rest. Using fixed size pools allows to lookup pool metadata in constant time (hashtable).

rainers commented:
> This is no-go, because it kills COW.

I also suspect that touching the pages to find the mark/noscan bits is not so great. My next try will move them back out of the object pages, but somehow make bits more condensed...

MartinNowak commented:
> I also suspect that touching the pages to find the mark/noscan bits is not so great. My next try will move them back out of the object pages, but somehow make bits more condensed...

Reading just suffers from general cache pollution and maybe TLB trashing, so indeed more compact is nicer.

rainers commented:
Rebased and moved byte array of bits out of the pages into a preallocated array. This is about as fast as master now, but isn't yet tweaked to reduce bit memory.

MartinNowak commented:
Will try to review this soon.
In the long-run I also want to try to get rid of pools and use a pageheap like tcmalloc.
http://goog-perftools.sourceforge.net/doc/tcmalloc.html

DmitryOlshansky commented:
> Will try to review this soon.

@MartinNowak any ETA ?

DmitryOlshansky commented:
Repinging @MartinNowak @rainers

rainers commented:
Rebased.

Geod24 commented:
Druntime have been merged into DMD. Please re-submit your PR to `dlang/dmd` repository.
